Download and cache package image contents in package manager

hasheddan opened the pull request:
<!--
Thank you for helping to improve Crossplane!

Please read through https://git.io/fj2m9 if this is your first time opening a
Crossplane pull request. Find us in https://slack.crossplane.io/messages/dev if
you need any help contributing.
-->

### Description of your changes

<!--
Briefly describe what this pull request does. Be sure to direct your reviewers'
attention to anything that needs special consideration.

We love pull requests that resolve an open Crossplane issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

-->

Updates the package manager to download static package images and cache them in a mounted volume. This PR will remain in draft while the following items are addressed:

- [x] Fix incompatible module dependencies (more info: https://github.com/Azure/go-autorest/issues/414) (this was fixed by removing a bunch of the docker machinery)
- [x] Complete unit test implementation (this is mostly done, but might follow up with some additional tests)
- [x] Allow for customizing volume mount in helm chart
- [x] Clean commit history

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

<!--
Before reviewers can be confident in the correctness of this pull request, it
needs to tested and shown to be correct. Briefly describe the testing that has
already been done or which is planned for this change.
-->

Manual testing details to follow.

[contribution process]: https://git.io/fj2m9

negz commented on pkg/controller/pkg/revision/imageback.go:
```diff
@@ -0,0 +1,171 @@
+		auth, err := k8schain.New(ctx, i.client, k8schain.Options{
+			Namespace:        i.namespace,
+			ImagePullSecrets: i.pullSecrets,
+		})
```
Would it make more sense to support passing in the entire auth keychain to this backend? Currently it seems like the the `Namespace`, `PullSecrets`, and `Client` options are all tightly coupled, in that you need to provide all three to make this work.

```go
func Keychain(k authn.Keychain) parser.BackendOption
```

hasheddan replied:
This gets difficult as the `namespace` is set when the controller is setup, but the `imagePullSecrets` get set for each individual `PackageRevision`. So while they are tightly coupled in usage, they are not in initialization.

negz replied:
I can see the rationale here, but it makes the ImageBackend API a little unusual and arguably tightly coupled to/designed around this particular consumer. I'd consider storing the parts that are fixed for the lifetime of the reconciler (e.g. `namespace`) in the Reconciler directly, and using them to init the image backend on each reconcile.

hasheddan replied:
Yeah the parser backends are strange in the regard that they can be set and then later overridden. It also is a powerful feature in terms of dropping new backends into something like a reconciler because, as can be seen in this PR, you can easily swap out a backend with very minimal changes. Also makes testing much easier. That was the original impetus for configuring constant things in the backend at controller instantiation.

However, I have updated to pass those things to the reconciler here as you suggested. Still feels a bit gross to me, but it serves its purpose well and I think we can circle back on this if needed in future.

negz commented on pkg/controller/pkg/revision/imageback.go:
```diff
@@ -0,0 +1,171 @@
+// ImageBackend is a backend for parser.
+type ImageBackend struct {
+	pkg         string
+	id          string
```
Should `pkg` and `id` be required arguments to `NewImageBackend`? They don't appear to be optional. e.g.

```go
func NewImageBackend(pkg, id string, bo ...parser.BackendOption) *ImageBackend
```

hasheddan replied:
Similarly, these both get set in the reconcile loop

negz commented on pkg/controller/pkg/revision/imageback.go:
```diff
@@ -0,0 +1,171 @@
+type ImageBackend struct {
+	pkg         string
+	id          string
+	cache       cache.Cache
```
Similarly, the `Cache` option appears to be required in practice at the moment. I'd lean toward keeping it optional and having the `ImageBackend` use a no-op cache that always returns an error on `Get` by default.

negz commented on pkg/controller/pkg/revision/reconciler.go:
```diff
@@ -224,12 +250,12 @@ func SetupConfigurationRevision(mgr ctrl.Manager, l logging.Logger, namespace st
 		WithParser(parser.New(metaScheme, objScheme)),
-		WithParserBackend(parser.NewPodLogBackend(parser.PodNamespace(namespace))),
-		WithLinter(NewPackageLinter([]PackageLinterFn{OneMeta}, []ObjectLinterFn{IsConfiguration}, []ObjectLinterFn{Or(IsXRD, IsComposition)})),
+		WithParserBackend(NewImageBackend(Client(clientset))),
```
Should this have a cache, etc?

hasheddan replied:
Yep :+1: was just testing one at a time and forgot to update

negz commented on pkg/controller/pkg/revision/imageback.go:
```diff
@@ -0,0 +1,171 @@
+}
+
+// NewImageBackend creates a new image backend.
+func NewImageBackend(bo ...parser.BackendOption) *ImageBackend {
```
What's the thinking behind accepting a set of backend options both when creating this backend and initialising it? Do we need to allow both?

hasheddan replied:
See earlier comments about things initialized at controller creation and things initialized on each reconcile.

negz commented on pkg/cache/cache.go:
```diff
@@ -0,0 +1,90 @@
+	if err != nil {
+		return err
+	}
+	cf, err := c.fs.Create(fmt.Sprintf(tarFmt, c.dir, id))
```
What does this file path look like in practice? I have a bit of trouble following the naming logic; do we include the digest in the file path to ensure a cache lookup won't return an older version of a retagged image?

hasheddan replied:
The tarball is has the same name as the revision, so it would look something like `provider-aws-<truncated-image-digest>`. The digest ensures we won't get false cache hits. One thing to note here is that the `Package` controller would continue to look for the image at the given tag. This is not so different than the existing model, but I like to point it out because in the case that someone uses a tag like `latest`, if a reconcile gets triggered for the `Package` controller it will create a new `PackageRevision` and in the case where a user has `revisionActivationPolicy: Active`, the package manager will _automatically_ update to the new revision. If someone has `revisionActivationPolicy: Manual` the new revision will get created but we will not upgrade to it until they manually mark it as `Active`. So the safest policy would be to always use specific tags and set `revisionActivationPolicy: Manual`. However, in practice, just using specific tags will likely be sufficient as underlying contents should not be changing in a well-versioned project.

hasheddan commented on pkg/cache/cache.go:
```diff
@@ -0,0 +1,90 @@
+limitations under the License.
+*/
+
+package cache
```
Already moved this to `image` pkg as this became even more gross when I started to build similar functionality in `crank` :+1:

hasheddan commented on cmd/crossplane/core/core.go:
```diff
@@ -35,13 +37,15 @@ import (
 func FromKingpin(cmd *kingpin.CmdClause) *Command {
 	c := &Command{Name: cmd.FullCommand()}
 	cmd.Flag("namespace", "Namespace used to unpack and run packages.").Short('n').Default("crossplane-system").OverrideDefaultFromEnvar("POD_NAMESPACE").StringVar(&c.Namespace)
+	cmd.Flag("cache-dir", "Directory used for caching package images.").Short('c').Default("/cache").OverrideDefaultFromEnvar("PACKAGE_CACHE_DIR").StringVar(&c.CacheDir)
```
I think I'll change env var to `CACHE_DIR` here. Also, I think have per-command env var may override the app level ones:

> DefaultEnvars configures all flags (that do not already have an associated envar) to use a default environment variable in the form "<app>_<flag>".

hasheddan commented on cmd/crank/main.go:
```diff
@@ -18,20 +18,34 @@ package main
 	ctx := kong.Parse(&cli,
 		kong.Name("crank"),
 		kong.Description("A tool for building platforms on Crossplane."),
+		kong.Bind(child),
```
This binds `child`, making it possible to pass it to command `Run` functions.

hasheddan commented on cmd/crank/main.go:
```diff
@@ -18,20 +18,34 @@ package main
+	strVal string
+}
+
+// strChild is a string value child argument.
+type strChild string
+
+func (c strChild) AfterApply(arg *childArg) error { // nolint:unparam
+	arg.strVal = string(c)
+	return nil
 }
```
Any time an arg or flag of type `strChild` is traversed in the node tree, `AfterApply` will be invoked. I am using this to propagate values from child to parent commands.

hasheddan commented on cmd/crank/push.go:
```diff
@@ -0,0 +1,84 @@
+}
+
+// Run runs the push cmd.
+func (c *pushCmd) Run(child *childArg) error {
```
Logic in the run function of a command with child node subcommands will be invoked after the subcommands are. This means that we can use common functionality in the parents and just use the child leaf nodes for value setting, etc.

hasheddan reviewed (commented):
@negz I have removed the [WIP] status but am keeping in draft so that we don't merge this and stomp on @muvaf work in #1798. Once I rebase on his changes and do a bit of refactoring then this should be ready for merge. I'll add full workflow to PR body for an understanding of how CLI works.

hasheddan commented on cmd/crank/build.go:
```diff
@@ -0,0 +1,110 @@
+}
+
+// Run runs the build cmd.
+func (c *buildCmd) Run(child *childArg) error {
```
Note: this does not currently lint the package while building, I would prefer to introduce that and the stand-alone `lint` command as well in a separate PR that includes switching to https://github.com/crossplane/crossplane-runtime/pull/211 since this PR is already so big.

hasheddan commented on cluster/charts/crossplane/templates/deployment.yaml:
```diff
@@ -44,4 +44,12 @@ spec:
-                fieldPath: metadata.namespace
\ No newline at end of file
+                fieldPath: metadata.namespace
+        volumeMounts:
+          - mountPath: /cache
+            name: package-cache
+      volumes:
+      - name: package-cache
+        emptyDir:
+          medium: {{ .Values.packageCache.medium }}
+          sizeLimit: {{ .Values.packageCache.sizeLimit }}
```
We probably want to be able to mount an existing volume here (that may already have some packages cached). That being said, I think it could also be a follow-up PR.

hasheddan commented:
@negz I believe I have addressed all feedback from our synchronous review today, plus quite a bit more in terms of testing and API cleaning (including adding linting to the build commands). I still anticipate further improvements will be made in the future in terms of how we structure and test `crank`, but I don't feel like we have any critical paths that are untested right now (besides my comment on the afero bug).

hasheddan commented on pkg/controller/pkg/revision/imageback_test.go:
```diff
@@ -0,0 +1,178 @@
+			},
+			want: errors.Wrap(&os.PathError{Op: "open", Path: xpkg.StreamFile, Err: syscall.ENOENT}, errOpenPackageStream),
+		},
+		// TODO(hasheddan): re-enable after https://github.com/spf13/afero/pull/268 is merged.
+		// "ErrEmptyImage": {
+		// 	reason: "Should return error if image is empty.",
+		// 	args: args{
+		// 		c: xpkg.NewNopCache(),
+		// 		f: &MockFetcher{
+		// 			MockFetch: func() (v1.Image, error) {
+		// 				return empty.Image, nil
+		// 			},
+		// 		},
+		// 		opts: []parser.BackendOption{Package("test/test:latest")},
+		// 	},
+		// 	want: errors.Wrap(&os.PathError{Op: "open", Path: "package.yaml", Err: syscall.ENOENT}, errOpenPackageStream),
+		// },
```
This test causes a panic because of the way the afero is handling tarfs initialization when an empty tar reader is passed. I have opened https://github.com/spf13/afero/pull/268 to fix. In practice, this would certainly be an edge case (i.e. totally empty package), but we obviously don't opportunity for panics in our code. I will work with afero maintainers to get this addressed prior to Crossplane release, or introduce a check and error early to avoid this execution branch.

negz commented on pkg/controller/pkg/revision/imageback.go:
```diff
@@ -0,0 +1,174 @@
+	return f, nil
+}
+
+// Fetcher is a tiny interface to allow for easier testing of the image backend.
```
Nit: While this is surely true (and would probably be true for 90% of the interfaces we use), I'd prefer that it described what the fetcher was intended to do; presumably to fetch the image with the given reference.

negz commented on pkg/controller/pkg/revision/reconciler.go:
```diff
@@ -121,52 +149,30 @@ func WithParserBackend(p parser.Backend) ReconcilerOption {
+	objects  Establisher
+	parser   parser.Parser
+	linter   parser.Linter
+	backend  parser.Backend
```
I wonder if `parser` is the right name for the crossplane-runtime package. It's not immediately obvious what it's a parser for (i.e. packages), and `parser.Parser` stutters. `xpkg` would make sense if it didn't conflict with the local package of that name. Perhaps `package.Parser`, `package.Linter`, and `package.Backend`?

hasheddan replied:
Yeah I struggled with this a bit. Similarly if it was named `package` we would have `package.Package`. I think `parser` actually makes sense for `Parser` (despite stuttering) and `Backend` (i.e. it the backend for a parser). `Linter` is linting the output of a parser, which would be a package so that is kind of strange. I guess we could name it `pkg`, which would make the general crossplane-runtime library `pkg` and the crossplane specific additions `xpkg`... If we make this change I will likely do so in follow-up PR

negz commented on pkg/xpkg/build.go:
```diff
@@ -0,0 +1,95 @@
+	_, err = io.Copy(buf, r)
+	if err != nil {
+		return nil, errors.Wrap(err, errCopyStream)
+	}
```
Nit: This (and the instance below) could be inlined to `if _, err := ...; err != nil`. (Feels like there should be a linter for this.)

negz commented on pkg/xpkg/find.go:
```diff
@@ -0,0 +1,60 @@
+	}
+	path := ""
+	for _, file := range files {
+		match, err := filepath.Match(XpkgMatchPattern, file.Name())
+		if err != nil {
+			return "", err
+		}
```
```suggestion
		// Match only returns an error if XpkgMatchPattern is malformed.
		match, _ := filepath.Match(XpkgMatchPattern, file.Name())  //nolint:errcheck
```

> The only possible returned error is ErrBadPattern, when pattern is malformed.

Usually I'd recommend using the `MustMatch` variant to panic here since the only way we'd hit an error is if the program was fundamentally broken, but none exists in this case. I suggest we simply don't handle this error.

negz commented on pkg/xpkg/name.go:
```diff
@@ -0,0 +1,94 @@
+	XpkgMatchPattern string = "*.xpkg"
+
+	// DefaultRegistry is the default registry for a Crossplane package.
+	DefaultRegistry string = "registry.upbound.io"
```
Is this used anywhere at the moment?

hasheddan replied:
Nope! It can be removed 👍

negz commented on pkg/xpkg/name.go:
```diff
@@ -0,0 +1,94 @@
+	XpkgExtension string = ".xpkg"
+
+	// XpkgMatchPattern is the match pattern for identifying compiled Crossplane packages.
+	XpkgMatchPattern string = "*.xpkg"
```
```suggestion
	XpkgMatchPattern string = "*" + XpkgExtension
```

Nit

negz reviewed (approved):
Awesome work @hasheddan! It's been a winding path, but I'm glad we landed here. This feels like a simplification. I've added a few comments but they're all nitpicks; feel free to ignore any you don't agree with or would prefer to defer, and merge at your leisure.

hasheddan commented on cmd/crank/build.go:
```diff
@@ -0,0 +1,144 @@
+	pkgName := child.name
+	if pkgName == "" {
+		metaPath := filepath.Join(root, xpkg.MetaFile)
+		pkgName, err = xpkg.ParseNameFromMeta(child.fs, metaPath)
```
@negz I believe `filepath.Join()` also runs `filepath.Clean()` on the returned result https://golang.org/pkg/path/filepath/#Join

hasheddan commented on cmd/crank/build.go:
```diff
@@ -0,0 +1,144 @@
+
+// Run runs the Configuration build cmd.
+func (c *buildConfigCmd) Run() error {
+	return nil
```
I actually discovered that they are still recognized as commands and run their hooks without the presence of the `Run()` method.. so that can just be eliminated :)
